File: rust/year2024/src/bin/day2.rs
use std::time::Instant;

use aoc_shared::{read_input, report_times};

fn is_safe(report: &Vec<isize>) -> bool {
    let ascending = report
        .windows(2)
        .all(|w| w[1] - w[0] >= 1 && w[1] - w[0] <= 3);
    let descending = report
        .windows(2)
        .all(|w| w[0] - w[1] >= 1 && w[0] - w[1] <= 3);

    ascending || descending
}

fn is_safe_benevolent(report: &Vec<isize>) -> bool {
    (0..report.len()).any(|i| {
        let mut reduced = report.clone();
        reduced.remove(i);
        is_safe(&reduced)
    })
}

fn main() {
    let start = Instant::now();

    // read and parse file
    let input = read_input("2024", "2", false).unwrap();
    let mut reports: Vec<Vec<isize>> = vec![];

    for line in input {
        let line = line.unwrap();
        if line.is_empty() {
            continue;
        }
        let mut report: Vec<isize> = vec![];
        for number in line.split_whitespace() {
            report.push(number.parse().unwrap());
        }
        reports.push(report);
    }
    let file_read_time = start.elapsed();

    // part 1
    println!(
        "{}",
        reports.iter().filter(|report| is_safe(&report)).count()
    );
    let part1_time = start.elapsed();

    // part 2
    println!(
        "{}",
        reports
            .iter()
            .filter(|report| is_safe_benevolent(&report))
            .count()
    );
    let part2_time = start.elapsed();

    // report times
    report_times(file_read_time, part1_time, part2_time);
}

File: python/src/year2025/day2.py
from time import time_ns

from regex import findall

from infra import read_input, report_times


def is_id_invalid(i: int) -> bool:
    id_string = str(i)
    if len(id_string) % 2 != 0:
        return False

    return id_string[: len(id_string) // 2] == id_string[len(id_string) // 2 :]


def is_id_invalid_2(number: int) -> bool:
    id_string = str(number)
    match = findall(r"^(.+)\1+$", id_string)
    return bool(match)


def day2() -> None:
    start = time_ns()

    # read and parse file
    inputs = read_input(2025, 2, example=False).strip().split(",")

    invalid_ids: set[int] = set()
    for string in inputs:
        begin, end = string.split("-")
        for i in range(int(begin), int(end) + 1):
            if i in invalid_ids:
                continue
            if is_id_invalid(i):
                invalid_ids.add(i)
    file_read_time = time_ns() - start

    # part 1
    print(sum(invalid_ids))
    part1_time = time_ns() - start

    # part 2
    invalid_ids: set[int] = set()
    for string in inputs:
        begin, end = string.split("-")
        for i in range(int(begin), int(end) + 1):
            if i in invalid_ids:
                continue
            if is_id_invalid_2(i):
                invalid_ids.add(i)
    print(sum(invalid_ids))
    part2_time = time_ns() - start

    # report times
    report_times(file_read_time, part1_time, part2_time)


if __name__ == "__main__":
    day2()
